### PR TITLE
Allow JSDoc casts of parenthesized expressions in JS

### LINKED ISSUE #5158: Consider parenthesized expressions with JSDoc to be a form of type assertions
To quote @weswigham on https://github.com/Microsoft/TypeScript/issues/4447#issuecomment-143099295:

> in [the] closure compiler you do inline casts with jsdoc comments like so:
> 
> ``` ts
> var x = /** @type {Foo.Bar} */ (y);
> ```

We should consider supporting this in Salsa.

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #5158

Based on #17176, since that changes some of the machinery that could have affected this.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -17731,6 +17735,20 @@ namespace ts {
             return type;
         }
 
+        function checkParenthesizedExpression(node: ParenthesizedExpression, checkMode?: CheckMode): Type {
+            if (isInJavaScriptFile(node)) {
+                if (node.jsDoc) {
```
Comment (sandersn) [inline src/compiler/checker.ts]: why not combine these predicates?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -17731,6 +17735,20 @@ namespace ts {
             return type;
         }
 
+        function checkParenthesizedExpression(node: ParenthesizedExpression, checkMode?: CheckMode): Type {
+            if (isInJavaScriptFile(node)) {
+                if (node.jsDoc) {
+                    const typecasts = flatten<JSDocTag>(map(node.jsDoc, doc => filter(doc.tags, tag => tag.kind === SyntaxKind.JSDocTypeTag)));
```
Comment (sandersn) [inline src/compiler/checker.ts]: can you use `flatMap` instead? `filter` should always return an array so you don't need the `if (isArray ...` check in `flatten`.

File: tests/cases/conformance/jsdoc/jsdocTypeTagCast.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+/** @type {string} */
+var s;
+
+var a = /** @type {*} */("" + 4);
+var s = "" + /** @type {*} */(4);
+
+class SomeBase {
+    constructor() {
+        this.p = 42;
```
Comment (sandersn) [inline tests/cases/conformance/jsdoc/jsdocTypeTagCast.ts]: These tests cases don't have type annotations. But based on your follow-up comment in the associated bug, the syntax 

```js
/** @type {number} */
this.p = 42;
```
should also work. I just want to confirm that this is not addressed in this PR.

**Comment by weswigham** [comment]
I'll reopen this against master (with feedback integrated) shortly.

**Pull request closed without merging**